Esta aula cobre os seguintes tópicos:

* Operações aritméticas com Python
* Resolver problemas em vários passos com variáveis
* Avaliar condições com Python
* Combinar condições e operadores lógicos
* Usar estilos de texto com Markdown

### Como executar o código

Esta aula é um [Jupyter notebook](https://jupyter.org) executável. Pode _correr_ esta aula e experimentar os exemplos de código de diferentes formas: *localmente no seu computador*, ou *utilizando um serviço online gratuito*.

#### Opção 1: Correr localmente no seu computador

Para correr localmente o código no seu computador, faça download do notebook e abra o ficheiro com uma aplicação ou ambiente de desenvolvimento suportado, por exemplo:
* Visual Studio Code: https://code.visualstudio.com/download
* Anaconda: https://www.anaconda.com/download
* Miniconda: https://docs.conda.io/projects/miniconda/en/latest/

Em qualquer das opções, as aplicações terão de suportar (nativamente ou através de extensões) o [Python](https://www.python.org) e os [Jupyter notebooks](https://jupyter.org), de forma a disponibilizar um ambiente de visualização e execução local com um kernel que execute o código Python contido no notebook.

#### Opção 2: Correr num serviço online gratuito

Para executar o notebook online, faça upload do notebook para o serviço da sua preferência, por exemplo:
* Google Colab: https://colab.google/
* Binder (com repositório GitHub): https://mybinder.org/
* Kaggle: https://www.kaggle.com/


>  **Jupyter Notebooks**: Esta aula é um [Jupyter notebook](https://jupyter.org) - um documento feito de _células_. Cada célula pode conter código escrito em Python ou explicações em português. Pode executar células de código e visualizar os resultados, e.g., números, mensagens, gráficos, tabelas, ficheiros, etc., instantaneamente no notebook. O Jupyter é uma plataforma poderosa para experimentação e análise. Não tenha medo de mexer no código ou estragar alguma coisa - aprenderá muito ao encontrar e corrigir erros. Pode utilizar a opção de menu "Kernel > Restart & Clear Output" (Kernel > Reiniciar e Limpar Saída) para limpar todas as saídas e recomeçar do início.

## Operações Aritméticas com Python

Vamos começar por usar o Python como uma calculadora. Podemos escrever e executar Python utilizando uma célula de código no Jupyter. 

> **Trabalhar com células**: Para criar uma nova célula no Jupyter, pode selecionar "Insert > Insert Cell Below" da barra de menu ou simplesmente clicar no botão "+" na toolbar. Pode também usar o atalho de teclado `Esc+B` para criar uma nova célula. Uma vez criada a célula, clique na mesma para a selecionar. Pode então alterar o tipo de célula para código ou markdown (texto), usando a opção de menu "Cell > Cell Type". Pode ainda usar os atalhos de teclado `Esc+Y` and `Esc+M`. Faça duplo clique sobre uma célula para editar o conteúdo da mesma. Para aplicar as alterações e executar a célula, utilize a opção de menu "Cell > Run Cells" ou clique no botão "Run" na toolbar, ou use simplesmente o atalho de teclado `Shift+Enter`. Pode ver uma lista completa de atalhos de teclado com a opção de menu "Help > Keyboard Shortcuts".

Execute as células de código abaixo para efetuar cálculos e ver os resultados. Tente alterar os números e voltar a executar as células alteradas para ver os resultados atualizados. Consegue adivinhar para que servem os operadores `//`, `%`, e `**`?

In [1]:
2 + 3 + 9

14

In [2]:
99 - 73

26

In [3]:
23.54 * -1432

-33709.28

In [4]:
100 / 7

14.285714285714286

In [5]:
100 // 7

14

In [6]:
100 % 7

2

In [7]:
5 ** 3

125

In [8]:
1000+1


1001

Como poderia esperar, operadores como `/` e `*` têm precedência sobre operadores como `+` e `-`, de acordo com as convenções da matemática. Pode usar parêntesis, i.e. `(` e `)`, para especificar a ordem pela qual as operações são efetuadas.

In [9]:
((2 + 5) * (17 - 3)) / (4 ** 3)

1.53125

O Python suporta os seguintes operadores aritméticos:

| Operador   | Operação          | Exemplo     | Resultado |
|------------|-------------------|-------------|-----------|
| `+`        | Adição            | `2 + 3`     | `5`       |
| `-`        | Subtração         | `3 - 2`     | `1`       |
| `*`        | Multiplicação     | `8 * 12`    | `96`      |
| `/`        | Divisão           | `100 / 7`   | `14.28..` |
| `//`       | Divisão inteira   | `100 // 7`  | `14`      |    
| `%`        | Módulo/Resto      | `100 % 7`   | `2`       |
| `**`       | Exponenciação     | `5 ** 3`    | `125`     |


Tente resolver alguns problemas simples desta página:
https://www.math-only-math.com/worksheet-on-word-problems-on-four-operations.html . 

Pde usar as células vazias abaixo e adicionar mais células se necessário.

## Resolver problemas em vários passos com variáveis

Vamos tentar resolver o seguinte problema com o Python: 

> Uma mercearia vende um saco de gelo por 1,25 € com um lucro de 20%. Se vender 500 sacos de gelo, qual será o lucro total?

Podemos listar a informação fornecida e gradualmente converter o problema numa expressão matemática que pode ser avaliada com o Python. 

*Custo do saco de gelo (€)* = 1.25

*Margem de lucro* = 20% = .2

*Lucro por saco (€)* = margem de lucro * custo do saco = .2 * 1.25

*Numero de sacos* = 500

*Lucro total* = número de sacos * lucro por saco = 500 * (.2 * 1.25)

In [10]:
500 * (.2 * 1.25)

125.0

Assim, a mercearia faz um lucro total de 125 €. Embora esta uma forma razoável de resolver um problema, o significado dos números não é totalmente claro ao olhar para o código. Podemos dar nomes a cada um dos números criando *variáveis* Python.

> **Variáveis**: Quando estamos a trabalhar com uma linguagem de programação com o Python, a informação é armazenada em *variáveis*. Pode pensar nas variáveis como caixas onde pode armazenar dados. Os dados armazenados numa variável são designados como o *valor* da variável.

In [11]:
custo_do_saco_de_gelo = 1.25

In [12]:
margem_de_lucro = .2

In [13]:
numero_de_sacos = 500

As variáveis `custo_do_saco_de_gelo`, `margem_de_lucro`, e `numero_de_sacos` contêm agora a informação fornecida no problema. Podemos verificar o valor de uma variável escrevendo o seu nome numa célula. Podemos combinar variáveis usando operações aritméticas para criar outras variáveis.

> **Code completion**: Ao escrever numa célula de código no Jupyter o nome de uma variável existente, escreva apenas os primeiros caracteres e prima a tecla `Tab` para fazer o autocomplete do nome da variável. Experimente escrever `mar` numa célula de código abaixo e prima `Tab` para fazer autocomplete para `margem_de_lucro`.

In [14]:
margem_de_lucro

0.2

In [15]:
lucro_por_saco = custo_do_saco_de_gelo * margem_de_lucro

In [16]:
lucro_por_saco

0.25

In [17]:
lucro_total = numero_de_sacos * lucro_por_saco

In [18]:
lucro_total

125.0

Se tentar visualizar o valor de uma variável que ainda não foi *definida*, i.e., que ainda não recebeu um valor usando a instrução de atribuição `nome_variavel = valor`, o Python exibe um erro.

In [19]:
lucro_liquido

NameError: name 'lucro_liquido' is not defined

Armazenar e manipular dados usando variáveis com nomes apropriados é uma excelente forma de explicar o que faz o seu código.

Vamos agora exibir o resultado do problema utilizando uma mensagem amigável. Podemos fazer isso com a *função* `print`.

> **Funções**: Uma função é um conjunto reutilizável de instruções. Recebe uma ou mais entradas (_inputs_), efetua certas operações, e frequentemente gera uma saída (_output_). O Python oferece muitas funções incorporadas como o `print` e também nos permite definir as nossas próprias funções.

In [20]:
print("A mercearia faz um lucro total de", lucro_total, "€")

A mercearia faz um lucro total de 125.0 €


> **`print`**: A função `print` é usada para exibir informação. Recebe um ou mais inputs, que podem ser texto (entre aspas, e.g., `"isto é texto"`), números, variáveis, expressões matemáticas, etc. Vamos aprender mais sobre variáveis e funções na próxima aula.

Criar uma célula de código para cada variável ou operação matemática pode tornar-se cansativo. Felizmente, o Jupyter permite escrever múltiplas linhas de código dentro de uma única célula. O resultado da última linha de código da célula será exibido como output.

Vamos rescrever a solução do nosso problema numa única célula.

In [21]:
# Guardar os dados de entrada em variáveis
custo_do_saco_de_gelo = 1.25
margem_de_lucro = .2
numero_de_sacos = 500

# Efetuar os cálculos necesários
lucro_por_saco = custo_do_saco_de_gelo * margem_de_lucro
lucro_total = numero_de_sacos * lucro_por_saco

# Exibir o resultado
print("A mercearia faz um lucro total de", lucro_total, "€")

A mercearia faz um lucro total de 125.0 €


Repare que estamos a usar o caracter `#` para adicionar *comentarios* ao longo do código. 

> **Comentários**: Os comentários e as linhas em branco são ignoradas durante a execução, mas são úteis para fornecer informação aos humanos (incluindo a si) sobre o que faz o código. Os comentários podem ser _inline_ (no final de uma linha de código), ou numa linha separada, ou ainda a ocupar múltiplas linhas.

Comentários inline e comentários de uma linha começam com `#`, enquanto os comentários de múltiplas linhas começam e terminam com três aspas, i.e. `"""`. Aqui ficam alguns exemplos de comentários de código:

In [22]:
meu_numero_favorito = 1 # um comentário inline

In [23]:
# Este comentário tem a sua própria linha
meu_numero_menos_favorito = 3

In [24]:
"""Este é um comentário de várias linhas.
Escreva tanto ou tão pouco quanto quiser.

Os comentários são realmente úteis para as pessoas que lêem o seu 
código, mas tente mantê-los curtos e diretos.

Além disso, se usar bons nomes de variáveis, então o seu código é
muitas vezes auto-explicativo, e pode até nem precisar de comentários!
"""
um_numero_neutro = 5



> **EXERCÍCIO**: Uma empresa de viagens quer levar um avião às Bahamas. O voo do avião custa 5000 euros. Até agora, 29 pessoas inscreveram-se para a viagem. Se a empresa cobra 200 euros por bilhete, qual é o lucro obtido pela empresa? Crie variáveis para cada quantidade numérica e utilize operações aritméticas apropriadas.

In [1]:
# Definir as variáveis
custo_voo = 5000
numero_pessoas = 29
preco_bilhete = 200

# Calcular o lucro
lucro = (preco_bilhete * numero_pessoas) - custo_voo

# Exibir o resultado
print("O lucro obtido pela empresa é de", lucro, "euros")


O lucro obtido pela empresa é de 800 euros


## Avaliar condições com Python

Para além das operações aritméticas, o Python também oferece várias operações para comparar números e variáveis.

| Operador    | Descrição                                                                  |
|-------------|----------------------------------------------------------------------------|
| `==`        | Verifica se os operandos são iguais                                        |
| `!=`        | Verifica se os operandos são não iguais                                    |
| `>`         | Verifica se o operando da esquerda é maior que o operando da direita       |
| `<`         | Verifica se o operando da esquerda é menor que o operando da direita       |
| `>=`        | Verifica se o operando da esquerda é maior ou igual ao operando da direita |
| `<=`        | Verifica se o operando da esquerda é menor ou igual ao operando da direita |

O resultado de uma operação de comparação é `True` ou `False` (note as maiúsculas `T` e `F`). Estas são keywords especiais em Python.
Vamos fazer algumas experiências com operadores de comparação.

In [25]:
meu_numero_favorito = 1
meu_numero_menos_favorito = 5
um_numero_neutro = 3

In [26]:
# Verificação de igualdade - True
meu_numero_favorito == 1

True

In [27]:
# Verificação de igualdade - False
meu_numero_favorito == meu_numero_menos_favorito

False

In [28]:
# Verificação de desigualdade - True
meu_numero_favorito != um_numero_neutro

True

In [29]:
# Verificação de desigualdade - False
um_numero_neutro != 3

False

In [30]:
# Verificação maior que - True
meu_numero_menos_favorito > um_numero_neutro

True

In [31]:
# Verificação maior que - False
meu_numero_favorito > meu_numero_menos_favorito

False

In [32]:
# Verificação menor que - True
meu_numero_favorito < 10

True

In [33]:
# Verificação menor que - False
meu_numero_menos_favorito < meu_numero_favorito

False

In [34]:
# Verificação maior ou igual que - True
meu_numero_favorito >= 1

True

In [35]:
# Verificação maior ou igual que - False
meu_numero_favorito >= 3

False

In [36]:
# Verificação menor ou igual que - True
3 + 6 <= 9

True

In [37]:
# Verificação menor ou igual que - False
meu_numero_favorito + um_numero_neutro <= 3

False

Tal como as operações aritméticas, o resultado de uma operação de comparação pode também ser guardado numa variável.

In [38]:
custo_do_saco_de_gelo = 1.25
saco_de_gelo_caro = custo_do_saco_de_gelo >= 10
print("O saco de gelo é caro?", saco_de_gelo_caro)

O saco de gelo é caro? False


## Combinar condições e operadores lógicos

Os operadores lógicos `and`, `or` e `not` operam sobre condições e valores `True` & `False` (também conhecidos como *booleanos*). O `and` e o `or` operam sobre duas condições, enquanto o `not` opera sobre uma condição única.

O operados `and` retorna `True` quando ambas as condições avaliam para `True`. Caso contrário, devolve `False`.

| `a`     | `b`    | `a and b` |
|---------|--------|-----------|
|  `True` | `True` | `True`    |
|  `True` | `False`| `False`   |
|  `False`| `True` | `False`   |
|  `False`| `False`| `False`   |


In [39]:
meu_numero_favorito

1

In [40]:
meu_numero_favorito > 0 and meu_numero_favorito <= 3

True

In [41]:
meu_numero_favorito < 0 and meu_numero_favorito <= 3

False

In [42]:
meu_numero_favorito > 0 and meu_numero_favorito >= 3

False

In [43]:
True and False

False

In [44]:
True and True

True

O operador `or` retorna `True` se pelo menos uma das condições avalia para `True`. Retorna `False` apenas se ambas as condições são `False`.

| `a`     | `b`    | `a or b`  |
|---------|--------|-----------|
|  `True` | `True` | `True`    |
|  `True` | `False`| `True`    |
|  `False`| `True` | `True`    |
|  `False`| `False`| `False`   |



In [45]:
um_numero_neutro = 3

In [46]:
um_numero_neutro == 3 or meu_numero_favorito < 0

True

In [47]:
um_numero_neutro != 3 or meu_numero_favorito < 0

False

In [48]:
meu_numero_favorito < 0 or True

True

In [49]:
False or False

False

O operador `not` retorna `False` se a condição for `True` e `True` se a condição for `False`.

In [50]:
not um_numero_neutro == 3

False

In [51]:
not meu_numero_favorito < 0

True

In [52]:
not False

True

In [53]:
not True

False

Os operadores lógicos podem ser combinados para formar condições complexas. Utilize parêntesis `(` e `)` para indicar a ordem pela qual os operadores lógicos devem ser aplicados.

In [54]:
(2 > 3 and 4 <= 5) or not (meu_numero_favorito < 0 and True)

True

In [55]:
not (True and 0 < 1) or (False and True)

False

Se não forem usados parêntesis, os operadores lógicos são aplicados da esquerda para a direita.

In [56]:
not True and 0 < 1 or False and True

False

Tome partido da natureza interativa dos Jupyter notebooks para fazer experiências com os operadores aritméticos, condicionais e lógicos em Python. Vamos aprender mais sobre variáveis e funções nas próximas aulas.

Há ainda um operador ternário em Python. Aprenda sobre ele aqui: https://data-flair.training/blogs/python-ternary-operator/

## Usar estilos de texto com Markdown

Adicionar explicações usando células de texto (como esta) é uma ótima forma de tornar o seu notebook informativo para outros leitores. Também é útil caso necessite de voltar a consultá-lo no futuro. Pode fazer duplo clique sobre uma célula de texto no Jupyter para a editar. Em modo de edição, irá reparar que o texto aparece um pouco diferente (por exemplo, o título da secção tem um prefixo `##`. Este texto foi
escrito utilizando Markdown, uma forma simples de adicionar estilos ao seu texto. Execute esta célula para ver o output sem os caracteres especiais. Pode alternar entre a fonte e o output para aplicar um estilo específico.

Por exemplo, pode usar um ou mais caracteres `#` no início de uma linha para criar headers de diferentes tamanhos:

# Header 1

## Header 2

### Header 3

#### Header 4

Para criar uma lista numerada ou não numerada (i.e. com _bullets_), comece simplesmente uma linha com `*` ou `1.`.

Uma lista de _bullets_:

* Item 1
* Item 2
* Item 3

Uma lista numerada:

1. Apple
2. Banana
3. Pineapple

Pode tornar um texto bold com `**`, e.g., **texto a bold**, ou torná-lo itálico com `*`, e.g., *texto em itálico.* Também pode criar links,  e.g., [um link](https://www.ips.pt). As imagens também são facilmente incorporadas:

![](https://raw.githubusercontent.com/davsimoes/mcde-pds/main/img/calvin_hobbes_600.png)

Outra característica muito interessante do Markdown é a possibilidade de incluir blocos de código. Note que os blocos de código dentro de células Markdown não podem ser executados.

```
# Efetuar os cálculos necesários
lucro_por_saco = custo_do_saco_de_gelo * margem_de_lucro
lucro_total = numero_de_sacos * lucro_por_saco

# Exibir o resultado
print("A mercearia faz um lucro total de", lucro_total, "€")

```

Pode aprender a sintaxe completa do Markdown aqui: https://learnxinyminutes.com/docs/markdown/

### Guardar o seu notebook

É muito importante guardar o seu trabalho com frequência. Pode continuar a trabalhar mais tarde num notebook que gravou anteriormente ou pode partilhá-lo com outras pessoas e permitir que executem o seu código.

## Questões para Revisão

Tente responder às seguintes questões para testar a sua compreensão sobre os tópicos cobertos neste notebook:

1. O que é um Jupyter notebook? 
2. Como é que pode adicionar uma nova célula de código por baixo de uma célula existente?
3. Como é que pode adicionar uma nova célula Markdown por baixo de uma célula existente?
4. Como é que pode converter uma célula de código numa célula Markdown ou vice-versa?
5. Como é que pode executar uma célula de código no Jupyter?
6. Quais são as diferentes operações aritméticas suportadas no Python?
7. Como é que pode efetuar operações aritméticas com o Python?
8. Qual é a diferença entre os operadores `/` e `//`?
9. Qual é a diferença entre os operadores `*` e `**`?
10. Qual é a ordem de precedência para os operadores aritméticos em Python?
11. Como é que pode especificar a ordem pela qual as operações aritméticas são efetuadas numa expressão com múltiplos operadores?
12. Como é que pode resolver um problema aritmético com vários passos no Python?
13. O que são variáveis? Porque é que elas são úteis?
14. Como é que pode criar uma variável em Python?
15. Qual é o operador de atribuição em Python?
16. Quais são as regras para dar um nome a uma variável em Python?
17. Como é que pode ver o valor de uma variável?
18. Como é que pode armazenar o resultado de uma expressão aritmética numa variável?
19. O que é que acontece se tentar aceder a uma variável que não foi definida?
20. Como é que pode exibir mensagens em Python?
21. Que tipo de entradas é que a função print aceita?
22. O que são comentários de código? Como é que podem ser úteis?
23. Quais são as diferentes formas de criar comentários em código Python?
24. Quais são as diferentes operações de comparação suportadas no Python?
25. Qual é o resultado de uma operação de comparação?
26. Qual é a diferença entre `=` e `==` em Python?
27. Quais são os operadores lógicos suportados no Python?
28. Qual é a diferença entre os operadores `and` e `or`?
29. Pode usar operadores de comparação e lógicos na mesma expressão?
30. Qual é a finalidade de usar parêntesis em expressões aritméticas ou lógicas?
31. O que é o Markdown? Por que é que é útil?
32. Como é que pode criar headings de diferentes dimensoes usando o Markdown?
33. Como é que pode criar listas numeradas e de bullets usando o Markdown?
34. Como é que pode criar texto bold ou itálico usando o Markdown?
35. Como é que pode incluir links e imagens em células Markdown?
36. Como é que pode incluir blocos de código em células Markdown?
37. É possível executar blocos de código em células Markdown?
38. Como pode gravar um Jupyter notebook?
39. Como pode visualizar e alterar um Jupyter notebook?
40. Onde é que pode aprender mais sobre operações aritméticas, condicionais e lógicas em Python?


## Solução para Exercício



> **EXERCÍCIO**: Uma empresa de viagens quer levar um avião às Bahamas. O voo do avião custa 5000 euros. Até agora, 29 pessoas inscreveram-se para a viagem. Se a empresa cobra 200 euros por bilhete, qual é o lucro obtido pela empresa? Crie variáveis para cada quantidade numérica e utilize operações aritméticas apropriadas.

In [57]:
custo_aviao = 5000
total_pessoas_inscritas = 29
custo_bilhete = 200

In [58]:
valor_recebido_por_empresa = total_pessoas_inscritas * custo_bilhete

In [59]:
lucro_total = valor_recebido_por_empresa - custo_aviao

In [60]:
lucro_total

800

A companhia de viagens fez um lucro de 800 €!

## Referências

Este notebook é uma adaptação traduzida do curso *<u>Data Analysis with Python: Zero to Pandas</u>* de AaKash N S / [Jovian.ai](https://jovian.ai)

Outras referências:
* McKinney, W., Python for Data Analysis, 3rd. Ed. O'Reilly. Versão online em https://wesmckinney.com/book/ 
* Documentação oficial do Python: https://docs.python.org/3/tutorial/index.html
* Tutorial Python do W3Schools: https://www.w3schools.com/python/
* Practical Python Programming: https://dabeaz-course.github.io/practical-python/Notes/Contents.html
* Jupyter Notebooks: https://docs.jupyter.org
* Markdown Reference: https://www.markdownguide.org
